src/ui/classic/xcbinputwindow.cpp

```diff
@@ -47,10 +47,22 @@ void XCBInputWindow::updatePosition(InputContext *inputContext) {
     }
     int x, y, h;
 
-    x = inputContext->cursorRect().left();
-    y = inputContext->cursorRect().top();
+    // add support of input panel offset here.
+    // TODO: RTL support.
+    auto &theme = parent_->theme();
+    int leftSW, rightSW, topSW, bottomSW, expectedWidth, expectedHeight;
+    leftSW = theme.inputPanel->shadowMargin->marginLeft.value();
+    rightSW = theme.inputPanel->shadowMargin->marginRight.value();
+    topSW = theme.inputPanel->shadowMargin->marginTop.value();
+    bottomSW = theme.inputPanel->shadowMargin->marginBottom.value();
+
+    x = inputContext->cursorRect().left() - leftSW;
+    y = inputContext->cursorRect().top() - topSW;
     h = inputContext->cursorRect().height();
 
+    expectedWidth = width() - leftSW - rightSW;
```

Comment:
expectedWidth is a little bit confusing, I'd prefer sth like actualWidth. Also, you need to make sure it's not less than zero.